channel/state.go

```diff
@@ -50,9 +51,14 @@ type (
 	}
 
 	// Data is the data of the application running in this app channel.
-	// Decoding happens with App.DecodeData.
+	//
+	// It is sent as binary data over the wire. For unmarshaling a data from
+	// binary representation, use App.NewData() to create an instance of Data
+	// specific to this application and then unmarshal using it.
 	Data interface {
-		perunio.Encoder
+		encoding.BinaryMarshaler
+		encoding.BinaryUnmarshaler
```

Comment:
Why need `Unmarshaler` on interface when we didn't need `Decoder` before?

Reply:
Because, earlier it was decoded by calling `App.DecodeData`, which instantiated an instance of Data and then decoded using it.

Now, `App` has method `NewData`, which just instantiates an instance of Data, which implements `encoding.BinaryUnmarshaler`. This Data can then be directly decoded by passing it to the `pkg.Decode` function.

Essentially, this change moves all notions of decoding to `pkg/io` package.